src/lib/comms/sol-mqtt.c

```diff
@@ -0,0 +1,579 @@
+            return NULL;
+        }
+    }
+
+    mqtt->connection_status = SOL_MQTT_DISCONNECTED;
+
+    r = mosquitto_connect_async(mqtt->mosq, host, port, mqtt->keepalive);
+    if (r != MOSQ_ERR_SUCCESS)
+        SOL_WRN("Unable to connect to %s:%d", host, port);
+
+    mqtt->socket_fd = mosquitto_socket(mqtt->mosq);
+    if (mqtt->socket_fd == -1) {
+        SOL_WRN("Unable to get socket file descriptor");
+        goto socket_error;
+    }
+
+    mqtt->socket_read = sol_fd_add(mqtt->socket_fd, SOL_FD_FLAGS_IN | SOL_FD_FLAGS_PRI,
+            sol_mqtt_event_loop, mqtt);
+    SOL_NULL_CHECK(mqtt->socket_read, NULL);
+
+    return mqtt;
+
+socket_error:
+    mosquitto_destroy(mqtt->mosq);
+error:
+    free(mqtt);
+    return NULL;
+}
+
+SOL_API int
+sol_mqtt_reconnect(struct sol_mqtt *mqtt)
+{
+    int r;
+
+    SOL_NULL_CHECK(mqtt, -EINVAL);
+
+    r = mosquitto_reconnect_async(mqtt->mosq);
+
+    if (r != MOSQ_ERR_SUCCESS) {
+        SOL_WRN("Unable to reconnect");
+        return -EINVAL;
+    }
+
+    if (mqtt->socket_read)
+        sol_fd_del(mqtt->socket_read);
+    if (mqtt->socket_write)
+        sol_fd_del(mqtt->socket_write);
+
+    mqtt->socket_read = NULL;
+    mqtt->socket_write = NULL;
+
+    mqtt->socket_fd = mosquitto_socket(mqtt->mosq);
+    if (mqtt->socket_fd == -1) {
+        SOL_WRN("Unable to get socket file descriptor");
+        return -EINVAL;
+    }
+
+    mqtt->socket_read = sol_fd_add(mqtt->socket_fd, SOL_FD_FLAGS_IN | SOL_FD_FLAGS_PRI,
+            sol_mqtt_event_loop, mqtt);
+
+    return 0;
+}
+
+SOL_API void
+sol_mqtt_disconnect(struct sol_mqtt *mqtt)
+{
+    CHECK_INIT();
+    SOL_NULL_CHECK(mqtt);
+
+    if (mqtt->socket_read)
+        sol_fd_del(mqtt->socket_read);
+    if (mqtt->socket_write)
+        sol_fd_del(mqtt->socket_write);
+
+    if (mqtt->connect_timeout)
+        sol_timeout_del(mqtt->connect_timeout);
+    if (mqtt->disconnect_timeout)
+        sol_timeout_del(mqtt->disconnect_timeout);
+    if (mqtt->publish_timeout)
+        sol_timeout_del(mqtt->publish_timeout);
+    if (mqtt->message_timeout)
+        sol_timeout_del(mqtt->message_timeout);
+    if (mqtt->subscribe_timeout)
+        sol_timeout_del(mqtt->subscribe_timeout);
+    if (mqtt->unsubscribe_timeout)
+        sol_timeout_del(mqtt->unsubscribe_timeout);
+
+    mosquitto_disconnect_callback_set(mqtt->mosq, NULL);
+    mosquitto_disconnect(mqtt->mosq);
+
+    mosquitto_destroy(mqtt->mosq);
+    free(mqtt);
+
+    sol_mqtt_shutdown();
+}
+
+SOL_API int
+sol_mqtt_get_connection_status(struct sol_mqtt *mqtt)
+{
+    SOL_NULL_CHECK(mqtt, -EINVAL);
+    return mqtt->connection_status;
+}
+
+SOL_API int
+sol_mqtt_publish(struct sol_mqtt *mqtt, struct sol_mqtt_message *message)
+{
+    int r;
+
+    CHECK_INIT(-EINVAL);
+    SOL_NULL_CHECK(mqtt, -EINVAL);
+    SOL_NULL_CHECK(message, -EINVAL);
+
+    r = mosquitto_publish(mqtt->mosq, &message->id, message->topic, message->payload_len, message->payload,
```

Comment:
topic is a NULL terminated string here. We allocate one extra byte for the \0 when the message is created.